File: fathersDay/Dsa/Day 33 (71. Greatest Common Divisor of Strings)/Solution.java
import java.math.BigInteger;

class Solution {
    public String GreatestCommonDivisor(String str1, String str2) {
        if (!(str1 + str2).equals(str2 + str1)) {
            return "";
        }
        int gcdLength = BigInteger.valueOf(str1.length()).gcd(BigInteger.valueOf(str2.length())).intValue();
        return str1.substring(0, gcdLength);
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        String str1 = "ABCABC";
        String str2 = "ABC";
        String result = solution.GreatestCommonDivisor(str1, str2);
        System.out.println(result); // Output: "ABC"
    }
}

File: fathersDay/Dsa/Day 33 (71. Greatest Common Divisor of Strings)/Solution.py
class Solution:
    def GreatestCommonDivisor(self, str1: str, str2: str) -> str:
        import math
        if str1 + str2 != str2 + str1:
            return ""
        gcd_len = math.gcd(len(str1), len(str2))
        return str1[:gcd_len]
# Example usage:
solution = Solution()
str1 = "ABCABC"
str2 = "ABC"
result = solution.GreatestCommonDivisor(str1, str2)
print(result)  # Output: "ABC"